File: DFS/Lint_619.java
// https://www.lintcode.com/problem/binary-tree-longest-consecutive-sequence-iii/description?_from=ladder&&fromId=1

// M1, resultType
/**
 * Definition for a multi tree node.
 * public class MultiTreeNode {
 *     int val;
 *     List<MultiTreeNode> children;
 *     MultiTreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    /**
     * @param root the root of k-ary tree
     * @return the length of the longest consecutive sequence path
     */
    class ResultType {
        int maxLen, inc, dec;
        public ResultType (int maxLen, int inc, int dec) {
            this.maxLen = maxLen;
            this.inc = inc;
            this.dec = dec;
        }
    }
    public int longestConsecutive3(MultiTreeNode root) {
        return helper(root).maxLen;
    }
    
    private ResultType helper(MultiTreeNode root) {
        if (root == null)
            return new ResultType(0, 0, 0);
        
        int inc = 1, dec = 1, maxLen = 1;
        for (MultiTreeNode child: root.children) {
            ResultType res_child = helper(child);
            if (res_child.maxLen != 0) {
                if (child.val + 1 == root.val) 
                    inc = Math.max(inc, res_child.inc + 1);
                    
                if (child.val - 1 == root.val)
                    dec = Math.max(dec, res_child.dec + 1);
                    
                maxLen = Math.max(maxLen, res_child.maxLen);
            }
        }
        
        maxLen = Math.max(maxLen, inc + dec - 1);
        
        return new ResultType(maxLen, inc, dec);
    }
}

File: 5 DFS_py/Lint_619.py
"""
Definition for a multi tree node.
class MultiTreeNode(object):
    def __init__(self, x):
        self.val = x
        children = [] # children is a list of MultiTreeNode
"""


class Solution:
    # @param {MultiTreeNode} root the root of k-ary tree
    # @return {int} the length of the longest consecutive sequence path
    def longestConsecutive3(self, root):
        up, down, max_len = self.helper(root)
        return max_len
    
    def helper(self, root):
        if root is None:
            return 0, 0, 0
        
        up_lst, down_lst, max_len_lst = [], [], []
        for c in root.children:
            c_up, c_down, c_max_len = self.helper(c)
            up_lst.append(c_up)
            down_lst.append(c_down)
            max_len_lst.append(c_max_len)
        
        up, down = 0, 0
        for i in range(len(root.children)):
            if root.val + 1 == root.children[i].val:
                down = max(down, down_lst[i] + 1)
            if root.val - 1 == root.children[i].val:
                up = max(up, up_lst[i] + 1)
        
        max_len = max(max_len_lst, default=0)
        max_len = max(down + up + 1, max_len)
        
        return up, down, max_len
